List refactor, Dropdown refactor, ComboBox refactor, Tabs refactor, Toolbar refactor, Tree component

heswell opened the pull request:
There is a lot of new content in this PR. It is broken into commits roughly by component, by filtering the files changed view by commit, this might make the task of navigating these changes a little more manageable. I have ordered the commits into what I think is the best order to approach this material:

1. List
2. Dropdown
3. ComboBox
4. Calendar, QueryInput etc converted to use new DropdownList
5. Tabs
6. Toolbar
7. Tree
8. misc

### List
There is a new implementation of List, much of whose functionality has been factored out into _common-hooks_. These hooks will also be shared with other components like Tree. These hooks are intended for components that manage the presentation of collections of similar items - think List, Tree, possibly TokenizedInput, Tabs etc. They support child items configured as _source_ or JSX _children_. (maybe they should collectively be called collection-hooks rather than common-hooks ?) . 

List work still to do:

- merge the List and VirtualizedList implementations
- add Text truncation detection / tooltip provision
- flip the assignment of forwardedRef - assign forwarded ref as root element ref, provide a ref prop to assign to imperativeHandle

We still have the original List implementation, now renamed ListDeprecated. This is because CascadingMenu and MultiSelectComboBox have yet to be refactored to use new List. ListDeprecated os not exposed outside the library.

### Dropdown
Dropdown has been split into 2 parts

- A generic DropdownBase component that is intended to be extended by any component that requires _dropdown-like_ functionality - a trigger component associated with a popup component. 
- Dropdown a dropdown that uses a DropdownButton and a List - this is a direct replacement for the existing Dropdown

Components that consume Dropdown, like Calendar, QueryInput andThemeEditor have been updated to use the new DropdownList.

Todo: A lot of code has been removed from Dropdown, it's possible that some code we needed has been lost. All original tests pass, I think  it's in the ADA area that we might have lost some functionality. ADA review should clarify.


### ComboBox
This has been reworked to use the new List and Dropdown. MultiselectCombobox has not yet been refactored. Because of this, we still have combobox-deprecated, the original combobox, which still contains MultiselectCombobox.

### Tabs
This is a refactoring of the existing Tabs. In particular, the responsive hooks that Tabs (and Toolbar) use have been heavily refactored. The TabPanel implementation is bare-bones - I anticipate that we will incorporate the Deck layout as soon as it is ready. Otherwise, Tabs is pretty much complete. 
Note: there is a single flaky test in the Tabs cypress suite. I think I know the issue. It does ArrowRight keyboard nav three times in succession. Focus on Arrow nav is done in a timeout. There's a race condition here when the ArrowKeys are invoked rapidly vis test script.

### Toolbar 
This is a work in progress. The responsive hooks have been heavily refactored. Alll the responsive behaviour should be   complete - work remaining is around the overflow panel and incorporation of skiplink.

### Tree
This is a partial implementation, included mainly to ensure that common hooks supported more than just List requirements.

origami-z commented:
Haven't looked at the code yet, played around the [deployed storybook](https://ffdfc1b8.uitk.pages.dev/?path=/story/lab-list--default) (mostly on List) and noticed below interaction differences
- When focused (either mouse or keyboard), there's no focus visible ring around list item
- When first mouse click to select an item (in the middle of list), then move the mouse away from list, then arrow down, highlighted item starts from the top instead of selected item
- (maybe related to above) after mouse click middle item and move mouse away, arrow up will not highlight first item which means there is no indication focus is on list or list item?
- When keyboard highlight last item (e.g. "End" / Fn + Right on mac), "Page Up" / Fn + Up will move highlight one more item than a full page (i.e. 11 items).
- Should "Type to select" feature stop propagation? E.g. "A" will close storybook panel

heswell commented:
Fixed the focusVisible and arrow navigation issues.

Toolkit 1 Page Down Page Up behaviour feels a little quirky too. List shows 9 items when it should show 10. Page down from initial state  highlights item 11, at the bottom of the scroll viewport, which scrolls by 2 rows.

I feel like correct behaviour should be that Page down page up scrolls by 1 'page' exactly, keeping highlighting in same relative position. So from initial view (items 1-10 visible) page down would result in items 11 - 20 visible. If no item was highlighted, then item 11 will now be visible . If item 2 was originally highlighted, then item 12 will now be highlighted.

 OR should pages overlap, so last row in viewport become first row after page down ?

origami-z commented:
After just looking  at List props and List stories
- I like the new ListItem can be “dumb” component, which doesn’t rely on consumer using hook
- Thumb up for unifying `default` as uncontrolled prefix
- Not quite sure about `selected` prop name given it can be perceived as a boolean type
- `itemGapSize` default to `0`, I thought we need 1px gap between items?
- `itemHeight` and `getItemHeight` just come to my attention this round that, is it confusing we have both and knowing when to use which (or both is used by mistake). Or is it part of virtualised v.s. not API?
- Will `selectionKeys` causing accessibility concerns? If we do want to offer customisation around this, is it possible by using `onSelectionChange` event & controlled props?
- Didn’t quite get how `ListItemGroup` actually works where it’s just returning null. Maybe slightly confusing for users who want to implement a custom one.

heswell commented:
> After just looking at List props and List stories
> 
> * Not quite sure about `selected` prop name given it can be perceived as a boolean type
See what you mean. I wanted to get away from the original selectedItem, which also seemed misleading as it can be an array of items if selectionStrategy is multi select

> * `itemGapSize` default to `0`, I thought we need 1px gap between items?
Will investigate, This started out as 1px then I switched to 0px (must have either been because of tk1 or figma)

> * `itemHeight` and `getItemHeight` just come to my attention this round that, is it confusing we have both and knowing when to use which (or both is used by mistake). Or is it part of virtualised v.s. not API?
Preserves tk1 API. getItemHeight is required if items can be different heights. Probably very rarely used, if ever.

> * Will `selectionKeys` causing accessibility concerns? If we do want to offer customisation around this, is it possible by using `onSelectionChange` event & controlled props?
Hm, not sure this is actually being used. WIll check. Was an idea I dropped.

> * Didn’t quite get how `ListItemGroup` actually works where it’s just returning null. Maybe slightly confusing for users who want to implement a custom one.
Hm, maybe confusing. Its being used as a 'marker' so I know to insert a Group. Will rethink

joshwooding commented:
On a GroupedList the Home key is broken

heswell commented on packages/lab/src/__tests__/__e2e__/calendar/Calendar.cy.tsx:
```diff
@@ -70,7 +70,9 @@ describe("GIVEN a Calendar component", () => {
           name: testDate.set("month", 4).format("MMM"),
-        }).realClick();
+        })
+          .realHover()
```
It's a bug in cypress (or the 'real' helpers). When realClick is invoked, mousemove, mouseover and hover events should fire before the click. Can't remember which one it was but one of these is not invoked and the navigation code depends on it (for tracking highlighted item). I found confirmation online that it is a known issue.

heswell commented on packages/lab/src/__tests__/__e2e__/dropdown/Dropdown.keyboardNavigation.cy.tsx:
```diff
@@ -12,61 +11,59 @@ import { Dropdown, MultiSelectDropdown } from "@jpmorganchase/uitk-lab";
-  describe(`Given a ${type}Dropdown component`, () => {
-    const testSource = ["Bar", "Foo", "Foo Bar", "Baz"];
-
+  const props = {
```
not sure I understand why, but I've removed them

heswell commented on packages/lab/src/__tests__/__e2e__/list/List.keyboardNavigation.cy.tsx:
```diff
@@ -0,0 +1,678 @@
+const ITEMS_PER_PAGE = 2;
+
+const expectOptionHightlightAt = (index: number) => {
+  cy.get(`#list-item-${index}`).should("have.class", HIGHLIGHTED, FOCUSVISIBLE);
```
I've added assertions for `highlighted`, `focusVisible` and `ariaSelected`. There are existing built-in assertions for `selected` and `disabled` which are not useful in this scenario. It would be nice to add our own check for `disabled" to check for the `uitkDisabled` utility class. Clashes with the built-in `disabled` assertion though. Maybe something like `uitk.disabled` ? (in which case we should perhaps go for  `uitk.selected` and `uitk.highlighted` as well ?

heswell commented on packages/lab/src/__tests__/__e2e__/tabs/Tabs.cy.tsx:
```diff
@@ -0,0 +1,703 @@
+      it("THEN first 4 tabs will be displayed, with overflow indicator", () => {
+        cy.mount(<SimpleTabstrip width={400} />);
+        cy.get(".uitkTabstrip").invoke("css", "width", "320px");
+        cy.get(".uitkTabstrip-inner > *")
```
the new assertions for things like focusVisible and ariaSelected help with readability here, though don't address the CSS selector issue. Problem is the tabs contain various content that aren't easy to target collectively with roles (tabs, add button, overflow button).
We could potentially create some cypress selectors specially for the `responsive` components (Tabs, Toolbar, Tooltray`, which could help abstract away some of this css ?

heswell commented on packages/lab/src/combo-box/useCombobox.ts:
```diff
@@ -0,0 +1,498 @@
+    disableAriaActiveDescendant,
+    disableHighlightOnFocus: true,
+    disableTypeToSelect: true,
+    label: "useComboBox",
```
It's only ever used in debugging , I'll remove them all

heswell commented on packages/lab/src/common-hooks/calcPreferredHeight.js:
```diff
@@ -0,0 +1,36 @@
+export const calcPreferredHeight = (props = {}) => {
```
this will disappear when Tree gets finished. This was originally used by List - that code has now been refactored into useListHeight. I suspect Tree can also be refactored to use useListHeight. -at which point we will probably want to rename it to useComponentHeight or useCollectionHeight or something similar.
I'll add a comment to this effect in calcPreferredHeight

heswell commented on packages/lab/src/common-hooks/key-code.ts:
```diff
@@ -0,0 +1,67 @@
+export const PageDown = "PageDown";
+export const Space = " ";
+
+const actionKeys = new Set(["Enter", "Delete", " "]);
```
what do you have against them ?

joshwooding replied:
They tend to be more memory intensive but maybe it's not a major issue?

heswell commented on packages/lab/src/common-hooks/key-code.ts:
```diff
@@ -0,0 +1,67 @@
+import React from "react";
```
renamed to keyUtils

heswell commented on packages/lab/src/common-hooks/use-auto-scroll.js:
```diff
@@ -0,0 +1 @@
+export const useAutoScroll = () => {};
```
deleted

heswell commented on packages/lab/src/common-hooks/use-drag-drop.js:
```diff
@@ -0,0 +1,340 @@
+          containerRef.current;
+        isScrollable.current = scrollSize > clientSize;
+
+        addEventListener("mousemove", preDragMouseMoveHandler, false);
```
yes

heswell commented on packages/lab/src/common-hooks/useViewportTracking.ts:
```diff
@@ -0,0 +1,173 @@
+  scrollIntoView: (item: CollectionItem<Item>) => void;
+}
+
+export const useViewportTracking = <Item>({
```
I struggled with the name of this one. Hard to make it really descriptive without it becoming a tongue-twister:

'useAutomaticallyScrollWhenUserNavigatesToAListItemOutsideTheViewport`

heswell commented on packages/lab/src/dropdown/Dropdown.css:
```diff
@@ -1,14 +1,21 @@
+}
+
+.uitkDropdown-popup-component {
+  max-height: 300px;
```
I honestly have no idea (removed)

heswell commented on packages/lab/src/query-input/internal/CategoryListItem.tsx:
```diff
@@ -33,19 +24,15 @@ export const CategoryListItem: FC<CategoryListItemProps> =
-
     return (
-      <ListItem item={category} onMouseMove={onMouseMove}>
+      <ListItem {...props} label={category?.name}>
```
It's always going to be present in this instance. Problem trying to type it more precisely here is that ListItemType is used to type all ListItems. Declarative list items may have no associated item, so its optional. 
I feel like it would complicate the typing quite a lot to accommodate this ?

heswell commented on packages/lab/src/toolbar/OverflowPanel.tsx:
```diff
@@ -0,0 +1,24 @@
+const withBaseName = makePrefixer("uitkOverflowPanel");
+
+export const OverflowPanel = forwardRef(function OverflowPanel(
+  { children, className, ...htmlAttributes }: OverflowPanelProps,
```
agree, according to our coding standards, it's htmlAttributes ;-)

https://confluence.prod.aws.jpmchase.net/confluence/display/JPMUITK/Coding+Standards

heswell commented on packages/lab/stories/tabs/tabs.stories.mdx:
```diff
@@ -0,0 +1,203 @@
+import { Meta, Story, Canvas } from "@storybook/addon-docs";
+import { useRef, useState } from "react";
```
they are being used, yes. Won't be necessary if the stories themselves get extracted out of the mdx file

heswell commented on packages/lab/stories/common-hooks/common-hooks.doc.stories.mdx:
```diff
@@ -0,0 +1,12 @@
+
+# useCollectionItems
+
+This hook manages a collection of Items, for components su
```
indeed

heswell commented on packages/lab/src/list/useList.ts:
```diff
@@ -1,569 +1,244 @@
+      : dataHook.data[highlightedIndex]?.id;
+
+  // We need this on reEntry for navigation hook to handle focus
+  lastSelection.current = selectionHook.selected;
```
This is a tricky one. I don't like the pattern I'm using and would welcome a better approach. I think it should still work with repeated renders a la React 18, though.

The problem I'm trying to solve is a circular dependency between the keyboard navigation hook and the selection hook. KeyboardNav needs to know the current selection, because we sometimes need to  begin navigation from the selected item.

But, selection only gets set after the keyboard nav hook is called and the selection hook needs the highlighted index which is returned from the keyboard nav hook. It needs this so that if Enter key is pressed, it selects the highlighted item.

So, we store the selection in a ref, initialised to the default value before the keyboard nav hook is called. It works because navigation always operates from the previous selection and navigation never changes selection. The selection hook must follow navigation, because selection must operate on the most up to fate highlighted index.

after both hooks have run, we set the selected ref to point to the current selection, ready for next render, whenever that happens.

heswell commented on packages/lab/src/responsive/useOverflowLayout.ts:
```diff
@@ -1,181 +1,184 @@
-  // console.log(JSON.stringify(managedItems, null, 2));
-  // console.groupEnd();
+  const { dispatch, version: collectionVersion } = collectionHook;
+  overflowItemsRef.current = collectionHook.data;
```
I think its safe here, but by all means explain to me why you think it isn't. The reason for doing this is to have a stable reference to the data to pass to nested hooks. There are callbacks that will operate on the data, but I don't need them to be regenerated every time the data changes - i.e I don't want to put colllectionHook.data directly in their dependency arrays - it's not necessary. Its ok if a render is aborted before completion, this ref  always gets set tot he correct value at the start of a new render.

heswell commented on packages/lab/stories/list.stories.tsx:
```diff
@@ -516,201 +471,160 @@ interface State {
-// );
 const ListPlaceholder = () => (
-  <span>TODO: Replace this string in a span with a ContentStatus</span>
+  <ContentStatus message="Did you hide it somewhere?" title="No source found" />
```
Not sure I understand - it is centered isn't it ?

heswell commented on packages/lab/stories/sample-components/CollapsibleToolbarItem.tsx:
```diff
@@ -0,0 +1,28 @@
+export const CollapsibleToolbarItem = ({
+  "data-collapsed": collapsed,
+  label = "Full Size",
+  labelCompact = "Compact",
```
yes, this example never got finished. I'll delete the unused stuff for now.
